### PR TITLE
Exposes B3SingleFormat after getting meticulous about parsing

### PR DESCRIPTION
**Pull request by adriancole**
This adds a huge amount of state checks to understand the source of
malform problems. This exposes B3SingleFormat methods so they can be
used in other code, such as w3c tracestate or JMS directly.

See #763

### CODE REVIEW

**Comment by adriancole** [comment]
fyi specifically for tracestate I think the parser needs to accept a (charsequence, pos, limit) or similar to avoid having to split strings.

and also here are the updated benchmark results so far:

```
B3SinglePropagationBenchmarks.extract:extract·p0.999                       sample             2.260            us/op
B3SinglePropagationBenchmarks.extract:·gc.alloc.rate.norm                  sample      15    80.010 ±  0.001    B/op
B3SinglePropagationBenchmarks.extract_malformed:extract_malformed·p0.999   sample             7.506            us/op
B3SinglePropagationBenchmarks.extract_malformed:·gc.alloc.rate.norm        sample      15     0.004 ±  0.001    B/op
B3SinglePropagationBenchmarks.extract_nothing:extract_nothing·p0.999       sample             0.149            us/op
B3SinglePropagationBenchmarks.extract_nothing:·gc.alloc.rate.norm          sample      15     0.001 ±  0.001    B/op
B3SinglePropagationBenchmarks.extract_unsampled:extract_unsampled·p0.999   sample             0.240            us/op
B3SinglePropagationBenchmarks.extract_unsampled:·gc.alloc.rate.norm        sample      15     0.002 ±  0.001    B/op
B3SinglePropagationBenchmarks.inject:inject·p0.999                         sample             9.232            us/op
B3SinglePropagationBenchmarks.inject:·gc.alloc.rate.norm                   sample      15   320.016 ±  0.003    B/op
```

**Comment by adriancole** [comment]
@drolando @ivantopo @nicmunroe fyi I will give a one week "quiet time" then brief you on all this stuff. That way you don't have to thrash through various optimizations :P

File: brave/src/main/java/brave/propagation/B3SingleFormat.java [hunk 1]
```diff
@@ -50,7 +74,7 @@ static String writeB3SingleFormat(TraceContext context) {
     return new String(result, 0, pos);
   }
 
-  static @Nullable TraceContextOrSamplingFlags maybeB3SingleFormat(String b3) {
+  @Nullable public static TraceContextOrSamplingFlags parseB3SingleFormat(String b3) {
```
Comment (adriancole) [inline brave/src/main/java/brave/propagation/B3SingleFormat.java]: this part, changing to CharSequence is one idea.

PS @llinder @michaelsembwever I think we should switch cassandra to use this single-header propagation format, since it is a good time anyway (as currently trace propagation is turned off in brave-cassandra!)

File: brave/src/main/java/brave/propagation/B3SingleFormat.java [hunk 2]
```diff
@@ -73,49 +97,83 @@ static String writeB3SingleFormat(TraceContext context) {
...
-      parentId = lenientLowerHexToUnsignedLong(b3, pos, pos + 16);
-      if (parentId == 0L) {
-        logger.log(FINE, "Invalid input: expected a 16 lower hex parent ID at offset {0}", pos);
+    if (length > pos) {
+      // If we are at this point, we have more than just traceId-spanId.
+      // If the sampling field is present, we'll have a delimiter 2 characters from now. Ex "-1"
+      // If it is absent, but a parent ID is (which is strange), we'll have at least 17 characters.
+      // Therefore, if we have less than two characters, the input is truncated.
+      if (length == pos + 1) {
+        logger.fine("Invalid input: truncated");
```
Comment (adriancole) [inline brave/src/main/java/brave/propagation/B3SingleFormat.java]: another idea I have is to make a base type which accepts a "parse error" operation. by default, this can go to JUL, but it could also go into a place where it can be a span tag.. for example propagation fail to parse reason.

**Comment by adriancole** [comment]
added ability to write byte array values. this is needed for cassandra (ByteBuffer) and kafka (byte[]) propagation

**Comment by adriancole** [comment]
incidentally grpc also uses byte arrays

**Comment by adriancole** [comment]
test coverage is 100% even edge cases.. will add the ability to parse ByteBuffer next. This will allow cassandra and kafka to not have to allocate arrays to read their data (ex ByteBuffer.wrap(array)  doesn't copy an array)
